### PR TITLE
[Feat/coordinator-nav] -> [dev]

### LINKED ISSUE #1: Extract navigation logic to a Coordinator/Flow Controller/Wireframe pattern
Currently view controllers present each other and "know" the way they are presented.
https://www.hackingwithswift.com/articles/71/how-to-use-the-coordinator-pattern-in-ios-apps
https://khanlou.com/2015/01/the-coordinator/

### PR DESCRIPTION
**Pull request by GGLozanov**
Implementation of a coordinator pattern for VC navigation. Might be better to use a static MainCoordinator instance from SceneDelegate but tutorial advised doing it with multiple weak reference instances. Nevertheless, the pattern works on emulator and has some small added functionality.

Requested review. @alexvidenov @MoridinBG.

### CODE REVIEW

File: GitHubExplorer/GitHubExplorer/Coordinators/Storyboarded.swift [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    
+    static func instantiate() -> Self // Self -> eventual type that conforms to this protocol
+}
+
+extension Storyboarded where Self: UIViewController { // only applicable to UIViewController
+    static func instantiate() -> Self {
+        // VC signature w/ NS: <AppName>.<VCName>
+        // => draw VC route and split by '.' character
+        
+        let vcFullName = NSStringFromClass(self)
```
Comment (MoridinBG) [inline GitHubExplorer/GitHubExplorer/Coordinators/Storyboarded.swift]: String(describing: self) should return only the class name without the module name

Comment (GGLozanov) [reply GitHubExplorer/GitHubExplorer/Coordinators/Storyboarded.swift]: Will fix it.

File: GitHubExplorer/GitHubExplorer/Coordinators/Coordinator.swift [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+//  GitHubExplorer
+//
+//  Created by ts51 on 6.07.20.
+//  Copyright © 2020 example. All rights reserved.
+//
+
+import UIKit
+
+protocol Coordinator {
+    var childCoordinators: [Coordinator] { get set }
```
Comment (alexvidenov) [inline GitHubExplorer/GitHubExplorer/Coordinators/Coordinator.swift:12]: We don't have an actual need for child coordinators (for now), so this var seems unnecessary.

Comment (GGLozanov) [reply GitHubExplorer/GitHubExplorer/Coordinators/Coordinator.swift:12]: Keeping it as per tutorial recommendation (and, in this case, it's left empty)

File: GitHubExplorer/GitHubExplorer/Coordinators/MainCoordinator.swift [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    var navigationController: UINavigationController
+    
+    init(navigationController: UINavigationController = UINavigationController()) {
+        self.navigationController = navigationController
+    }
+    
+    // FIXME: Generalise all these methods (important)
+    // and avoid repitition
+    
+    func navigateToRoot(shouldAnimate: Bool = true) {
```
Comment (MoridinBG) [inline GitHubExplorer/GitHubExplorer/Coordinators/MainCoordinator.swift]: shouldAnimate should be of concern only of the coordinator and not the caller

**Review by MoridinBG — APPROVED** [review]

**Pull request merged**